build.rs

```diff
@@ -271,8 +275,78 @@ const MSVC: &'static str = "msvc";
 const MSVC_OBJ_OPT: &'static str = "/Fo";
 const MSVC_OBJ_EXT: &'static str = "obj";
 
+const GENERATED_VERSIONED_HEADER: &str = "generated_versioned.h";
+const GENERATED_YASM_VERSIONED_HEADER: &str = "yasm_versioned.h";
+const GENERATED_VERSIONED_MACRO: &str = "VERSIONED";
+const GENERATED_INCLUDE_DIR: &str = "generated_headers";
+
+fn generated_include_dir() -> PathBuf {
+    PathBuf::from(::std::env::var("OUT_DIR").unwrap()).join(GENERATED_INCLUDE_DIR)
+}
+
+fn custom_versioned_symbols_header() -> PathBuf {
+    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("include").join("versioned.h")
+}
+
+fn yasm_versioned_symbols_header() -> PathBuf {
+    PathBuf::from(::std::env::var("OUT_DIR").unwrap())
+        .join(GENERATED_INCLUDE_DIR)
+        .join(GENERATED_YASM_VERSIONED_HEADER)
+}
+
+// Take the versioned header in `from`, replace #define with %define, preprocess
+// it with the C compiler, write it out to `yasm_versioned_symbols_header()`.
+fn write_yasm_versioned_header(from: &Path) -> io::Result<()> {
+    use std::fs::File;
+    use std::io::{BufReader, BufRead, Write, Cursor};
+
+    // Create the data for the intermediate file.
+    let mut new_data = String::new();
+    for line in BufReader::new(File::open(from)?).lines() {
+        use std::fmt::Write;
+        let line = line?;
+        if line.starts_with("#define") {
+            write!(new_data, "%{}\n", &line[1..]).expect("write to str");
+        } else {
+            write!(new_data, "{}\n", line).expect("write to str");
+        }
+    }
+
+    // Write the data to the intermediate file.
+    let output = yasm_versioned_symbols_header();
+    let intermediate = output.with_extension("in.h");
+    File::create(&intermediate)?.write_all(new_data.as_bytes())?;
+
+    // Expand the intermedia file.
+    let expanded = cc::Build::new()
+        .file(&intermediate)
+        .include(&generated_include_dir())
+        .expand();
+
+    // Write a header suitable for prepended to yasm files.
+    let mut out_file = File::create(&output)?;
+    for line in Cursor::new(expanded).lines() {
+        let line = line?;
+        if !line.starts_with('#') {
+            write!(out_file, "{}\n", line)?;
+        }
+    }
+
+    Ok(())
+}
+
+fn write_versioned_headers() {
+    let _ = write_versioned_header(&generated_include_dir(),
+            GENERATED_VERSIONED_HEADER, GENERATED_VERSIONED_MACRO)
+        .expect("generated versioned header file");
+
+    write_yasm_versioned_header(&custom_versioned_symbols_header())
+        .expect("generated yasm header file");
+}
 
 fn main() {
+    write_versioned_headers();
```

Comment:
How does this work w.r.t. incremental compilation?

1. How do we ensure all the C/asm code is recompiled when the versioned headers are changed? 
2. How do we ensure that the C/asm code isn't recompiled more than necessary?

Reply:
Currently, the generated `generated_versioned.h` and `yasm_versioned.h` headers are written out each time. Since these aren't tracked in `RING_INCLUDES`, they shouldn't cause a recompilation. The only header that should cause a recompilation is the manually written `versioned.h` header. This header is included in `RING_INCLUDES`, so it should trigger recompilation correctly. Thus, I don't think we need to do anything special here to get the right thing to happen. But if I've misunderstood something, do let me know.

The only other optimization I see is to only regenerate the generated header files if `versioned.h` is updated.

Reply:
generated_version.h will change whenever the *ring* version number is bumped, right? In that case, everything that includes it, directly or indirectly, needs to be recompiled, so that all the symbols get reversioned right? If so then generated_versioned.h. and yasm_versioned would need to be in `RING_INCLUDES` or we'd need equivalent logic.

Also, I think we need to have logic that builds these generated header files, then compares the contents wit the contents of any existing header files, and only overwrites the file if there is a difference. That way the time stamp of the files won't change spuriously and we wouldn't trigger recompilation of all the C code each time.

Reply:
So I'm not 100% certain how `native_versioning` works but shouldn't it be able to just compile the C files normally and then change the symbols before it links them with the Rust code? This wouldn't be as simple as running CC again on the same set of files with some changed, but in theory it should be fine.

Reply:
@clarcharr Yes, that's also a possible solution. That still requires you to know each symbol you want to change, however, which means you still have to keep a list somewhere. And, what's more, it also requires an `objcopy`-like utility to be present. Finding/invoking the utility in a cross-platform and cross-target way can be a challenge.

Reply:
@SergioBenitez you wouldn't need to keep a list; any symbol that's not linked to a dynamic library should be renamed, right?

Reply:
Yeah, that's true; all public symbols. Still, I think the other concern, requiring an `objcopy`-like utility, is the real issue, and the main reason why I didn't go this route. Note that such a utility would need to exist on the host for the _target_. I'm not sure how reasonable a requirement this is. We don't have any library like `cc` for `objcopy`, however, so we'd effectively need to write one.

It's a much more complex solution than what we have now. The current solution never recompiles or reads over binaries; it just compiles once with the symbols redefined.